languages/python/oso/oso/oso.py

```diff
@@ -45,17 +73,76 @@ def is_allowed(self, actor, action, resource) -> bool:
+        this method returns ``None``. If the action is not permitted by the
+        policy, this method will raise an error.
+
+        The error raised by this method depends on whether the actor can perform
+        the ``"read"`` action on the resource. If they cannot read the resource,
+        then a ``NotFound`` error is raised. Otherwise, a ``ForbiddenError`` is
+        raised.
+
+        :param actor: The actor performing the request.
+        :param action: The action the actor is attempting to perform.
+        :param resource: The resource being accessed.
+
+        :param check_read: If set to ``False``, a ``ForbiddenError`` is always
+            thrown on authorization failures, regardless of whether the actor can
+            read the resource. Default is ``True``.
+        :type check_read: bool
+
+        """
+        if self.query_rule_once("allow", actor, action, resource):
+            return
+
+        is_not_found = False
+        if check_read:
+            if action == self.read_action:
+                is_not_found = True
+            elif not self.query_rule_once("allow", actor, self.read_action, resource):
+                is_not_found = True
+        raise self.not_found_error() if is_not_found else self.forbidden_error()
+
+    def authorize_request(self, actor, request):
+        """Ensure that ``actor`` is allowed to send ``request`` to the server.
+
+        Checks the ``allow_request`` rule of a policy.
+
+        If the request is permitted with an ``allow_request`` rule in the
+        policy, then this method returns ``None``. Otherwise, this method raises
+        a ``ForbiddenError``.
+
+        :param actor: The actor performing the request.
+        :param request: An object representing the request that was sent by the
+            actor.
+        """
+        if not self.query_rule_once("allow_request", actor, request):
+            raise self.forbidden_error()
+
+    def authorized_actions(self, actor, resource, allow_wildcard=False) -> List[Any]:
+        """Determine the actions ``actor`` is allowed to take on ``resource``.
+
         Collects all actions allowed by allow rules in the Polar policy for the
         given combination of actor and resource.
 
+        Identical to ``Oso.get_allowed_actions``.
+
         :param actor: The actor for whom to collect allowed actions
 
         :param resource: The resource being accessed
 
         :param allow_wildcard: Flag to determine behavior if the policy \
-        includes a wildcard action. E.g., a rule allowing any action: \
-        ``allow(_actor, _action, _resource)``. If ``True``, the method will \
-        return ``["*"]``, if ``False``, the method will raise an exception.
+            includes a wildcard action. E.g., a rule allowing any action: \
+            ``allow(_actor, _action, _resource)``. If ``True``, the method will \
+            return ``["*"]``, if ``False``, the method will raise an exception.
```

Comment:
Maybe add an explanation of why the exception is raised? Something like "to avoid inadvertently indicating that all actions are permitted"... though now that I type that out it's really just alerting users to an issue with their policy, isn't it?

ETA: I see the explanation in the error message — maybe just align this description with that.